examples/run_language_modeling.py

```diff
@@ -144,8 +146,86 @@ def __getitem__(self, i):
         return torch.tensor(self.examples[i], dtype=torch.long)
 
 
+class LazyLineByLineTextDataset(Dataset):
+    """
+    Credit: @bramvanroy for this linecache implementation.
+    """
+
+    def __init__(self, file_path):
+        self.file_path = file_path
+        self.num_entries = self._get_n_lines(self.file_path)
+
+    @staticmethod
+    def _get_n_lines(file_path):
+        with Path(file_path).resolve().open(encoding="utf-8") as fhin:
+            for line_idx, _ in enumerate(fhin, 1):
+                pass
+
+        return line_idx
```

Comment:
I know that I suggested this function, but I found that it's still relatively slow. A faster approach is to just read the bytes and then count the number of new line characters. This supposes that ain input file is correctly formed.

```python
    @staticmethod
    def _get_n_lines(fin):
        # borrowed from https://stackoverflow.com/a/9631635/1150683
        def blocks(files, size=65536):
            while True:
                b = files.read(size)
                if not b:
                    break
                yield b

        with open(fin, encoding='utf-8') as fhin:
            n_lines = sum(bl.count("\n") for bl in blocks(fhin))
        return n_lines
```

Reply:
Here's markdown export of a notebook comparing a few methods. This new suggestion does indeed look to be the fastest. Increasing block size seems to help, so could expose as an argument of script, but probably not worth it.

```python
def _get_n_lines1(fin, size=65536):
    # borrowed from https://stackoverflow.com/a/9631635/1150683
    def blocks(files):
        while True:
            b = files.read(size)
            if not b:
                break
            yield b

    with open(fin, encoding='utf-8') as fhin:
        n_lines = sum(bl.count("\n") for bl in blocks(fhin))
    return n_lines
```


```python
from pathlib import Path 

def _get_n_lines2(file_path):
    with Path(file_path).resolve().open(encoding="utf-8") as fhin:
        for line_idx, _ in enumerate(fhin, 1):
            pass

    return line_idx

```


```python
def _get_n_lines3(file_path):
    return sum(1 for line in open(file_path, 'rb'))

```


```python
fn = 'shakespeare.txt'
```


```python
_get_n_lines1(fn), _get_n_lines2(fn), _get_n_lines3(fn)
```




    (124456, 124456, 124456)




```python
%%timeit
_get_n_lines1(fn)
```

    21.9 ms ± 1.26 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)



```python
%%timeit
_get_n_lines1(fn, size=32768)
```

    23.3 ms ± 920 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)



```python
%%timeit
_get_n_lines1(fn, size=131072)
```

    19.8 ms ± 235 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)



```python
%%timeit
_get_n_lines2(fn)
```

    71.1 ms ± 1.05 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)



```python
%%timeit
_get_n_lines3(fn)
```

    49.5 ms ± 469 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

Reply:
IIRC 65536 is the max number of bytes that 32-bit Python can read at a time, so I wouldn't change it. It is fastest as is, it seems, so it should be sufficient.